### DISCUSSION TITLE
FileUplink / Downlink via RF handling out-of-order or missing file packets

### DISCUSSION THREAD

**Discussion by ethancheez** [nasa/fprime#3063] [Q&A] [answered]
Does The FileUplink and FileDownlink components handle missing file packets?

Transferring files via RF often leads to packet loss, resulting in missing sections of a file. Is there any automated scheme for the ground to tell the flight OBC to retransmit those missing packets?

For example, I attempted downlinking a file greater 100 bytes from a flight OBC to the ground. I only received the first 100 bytes, and the rest was lost because the file packets were out of order. 
(P.S. I chose an arbitrary number of 100 bytes as an example)

**Comment by Joshua-Anderson** [comment, marked as answer]
Missed file chunks will be skipped and results in empty holes in the file. File uplink will emit a warning EVR if it sees out of order packets, or if the received file doesn't match the ground provided checksum.

Right now fprime's file transfer protocol is unreliable, so there's no way to automatically retry lost file packets. Manual re-transfers of lost file chunks can be commanded, but it's a bit painful since you have to do the math of what bytes are missing by hand.

There's some longer term work to implement the CCSDS file transfer protocols in fprime https://github.com/nasa/fprime/issues/2768 and Class 2 CCSDS file transfers are reliable file transfers.
